ui/src/controller/vsync_controller.ts

```diff
@@ -0,0 +1,121 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {assertExists} from '../base/logging';
+import {Actions} from '../common/actions';
+import {Engine} from '../common/engine';
+import {LONG, NUM} from '../common/query_result';
+import {Span} from '../common/time';
+import {globals} from '../frontend/globals';
+import {Controller} from './controller';
+
+export interface VsyncControllerArgs {
+  engine: Engine;
+}
+
+export class VsyncController extends Controller<'init'|'ready'> {
+  private engine: Engine;
+  private trackId?: number;
+  private timeSpan?: Span<bigint>;
+
+  constructor(args: VsyncControllerArgs) {
+    super('init');
+
+    this.engine = args.engine;
+  }
+
+  onDestroy() {
+    globals.dispatch(Actions.setVsyncData({}));
+  }
+
+  run() {
+    switch (this.state) {
+      case 'init':
+        this.getVsyncTrack().then((trackId) => this.trackId = trackId);
+        this.setState('ready');
+        break;
+      case 'ready':
+        if (this.trackId === undefined) {
+          // Still waiting for the track ID
+          return;
+        }
+        const visibleTimeSpan = globals.stateVisibleTime();
+        if (this.timeSpan === undefined ||
+            !this.timeSpan.equals(visibleTimeSpan)) {
+          this.timeSpan = visibleTimeSpan;
+          this.getVsyncData(this.trackId, this.timeSpan);
+        }
+        break;
+      default:
+        throw new Error(`Unexpected state ${this.state}`);
+    }
+  }
+
+  async getVsyncTrack(): Promise<number|undefined> {
+    // Determine the track ID of the SurfaceFlinger VSYNC-app counter, if
+    // it exists.
+    const result = await this.engine.query(`
+      select process_counter_track.id as trackId
+      from process
+        join process_counter_track using (upid)
+      where process.name like '%/surfaceflinger' and
+        process_counter_track.name='VSYNC-app'
+      limit 1;
+    `);
+    if (result.numRows() < 1) {
+      return undefined;
+    }
+
+    const trackId = result.firstRow({'trackId': NUM}).trackId;
+    return trackId;
+  }
+
+  async getVsyncData(trackId: number, timeSpan: Span<bigint>) {
+    // Get at least two changes of the counter in the currently
+    // visible timespan of the trace
+    const result = await this.engine.query(`
+      select ts, value from (
+        select row_number() over (order by ts) as rn, ts, value
+        from counter
+        where track_id = ${trackId} and ts >= ${timeSpan.start}
+        order by ts)
+      where rn <= 2 or ts <= ${timeSpan.end};
```

Comment:
Also, should this not be an `and` instead of `or`? Don't you want to find two or less rows that are within the given timespan?

Reply:
> Also, should this not be an `and` instead of `or`? Don't you want to find two or less rows that are within the given timespan?

No, we want _at least_ two results for convenience of the calculations when rendering, but if there are more in the currently visible span of the trace, then we want all of them. The "at least two" may include one or more outside of the currently visible span. Comments in the code may help.

> This is {inclusive,inclusive}. Is that intentional?

It's not clear what interval you're referring to. There is no interval comparison that I can see in the query. Besides, the rules for SQL queries are typically totally different to other programming conventions 😉  It's about as esoteric and obtuse as it gets in code.

Reply:
Ah, I see. That's the range that we actually present in the UI so I prefer to be inclusive on both ends 😀